lecture_3/03_project/calculator/internal/engine/calculatorEngine.go

```diff
@@ -0,0 +1,62 @@
+package engine
+
+import (
+	"context"
+	"log"
+)
+
+type Calculator struct {
+	consumer  Consumer
+	handler   Handler
+	publisher Publisher
+}
+
+// New creates and returns a new Calculator.
+func New(consumer Consumer, handler Handler, publisher Publisher) *Calculator {
+	return &Calculator{
+		consumer:  consumer,
+		handler:   handler,
+		publisher: publisher,
+	}
+}
+
+// Start will run the calculatorEngine.
+func (c *Calculator) Start(ctx context.Context) {
+	err := c.processBetsFromController(ctx)
+	if err != nil {
+		log.Println("Engine failed to process bets received from controller:", err)
+		return
+	}
+
+	err = c.processEventUpdates(ctx)
+	if err != nil {
+		log.Println("Engine failed to process event updates:", err)
+		return
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Calculator) processBetsFromController(ctx context.Context) error {
+	consumedBetsReceived, err := c.consumer.ConsumeBetsFromController(ctx)
+	if err != nil {
+		return err
+	}
+
+	resultingBets := c.handler.HandleBetsFromController(ctx, consumedBetsReceived)
+	c.publisher.PublishCalculatedBets(ctx, resultingBets)
```

Comment:
This is "dangerous" as you may end up with an infinite loop between Controller and Calculator, where one is always writing to `bets` queue and the other is always writing to `bets-calculated` queue. Calculator shouldn't produce to `bets-calculated` unless it has some calculation to share with the rest of the system (which is not the case if it just stored a new bet into its local storage).
The most accurate approach would be to return from `HandleBetsFromController` method only those bets for which calculation was indeed made and produce only those bets (e.g. if the bet arrived after the according event update). However, in our simplified betting platform it is perfectly fine just to not produce anything into `bets-calculated` when handling messages from `bets` queue. Let me know if this doesn't make sense to you.